utils/misc/exponential.go

```diff
@@ -0,0 +1,31 @@
+package misc
+
+import (
+	"golang.org/x/exp/constraints"
+)
+
+type Number interface {
+	constraints.Integer | constraints.Float
+}
+
+// ExponentialNumber is a simple exponentially increasing number.
+type ExponentialNumber[T Number] struct {
+	value T
+}
+
+// Reset resets the number to zero.
+func (expo *ExponentialNumber[T]) Reset() {
+	expo.value = 0
+}
+
+// Next returns the next number, which is the previous one multiplied by 2, always abiding by the min and max provided.
+func (expo *ExponentialNumber[T]) Next(min, max T) T {
+	expo.value *= 2
+	if expo.value > max {
+		expo.value = max
+	}
+	if expo.value < min {
```

Comment:
With <=1 the function will always return `min` e.g.:

```go
var exp misc.ExponentialNumber[float64]
t.Log(exp.Next(0.9, 1)) // prints 0.9 since 0.9^2 = 0.81 which is less than min
```

I'm pointing it out only because it is a generic component and it can be used by anyone. A power of a <1 is still possible, it's just that it decreases.
Given the intended use for this, perhaps we should complain if that is attempted. Either way it shouldn't be a big deal.

Reply:
The algorithm is not raising the previous number to any power, it is merely multiplying the previous number by 2. Adding another test case using the scenario you mentioned for clarity
```
	t.Run("exponential float", func(t *testing.T) {
		var exp misc.ExponentialNumber[float64]
		require.EqualValues(t, 0.9, exp.Next(0.9, 5))
		require.EqualValues(t, 1.8, exp.Next(0.9, 5))
		require.EqualValues(t, 3.6, exp.Next(0.9, 5))
		require.EqualValues(t, 5, exp.Next(0.9, 5))
		require.EqualValues(t, 5, exp.Next(0.9, 5))
		exp.Reset()
		require.EqualValues(t, 0.9, exp.Next(0.9, 5))
	})
```